### PR TITLE
Spark - Implement FunctionCatalog and Truncate

### PR DESCRIPTION
**Pull request by kbendick**
Implements `FunctionCatalog` for Spark 3.3 and implements all variants of `Truncate`.

## FunctionCatalog

This allows users of `SparkCatalog` and `SparkSessionCatalog` to use `truncate` without having to register it as a UDF.

All Iceberg functions that we register into the function catalog are accessible when used with an Iceberg spark catalog and:

1. No namespace is referenced - the storage partitioned joins implementation requires this.
e.g. `my_catalog.truncate(width, value)`. 
**Note** - Using `truncate(width, value)` typically does not work, as Spark adds the namespace to the call. `system.truncate` should be preferred.
2. The `system` namespace is referenced, to match called procedure syntax. **Note** this only works right now with the `SparkCatalog`, as the `SparkSessionCatalog` has logic in Spark [to verify the namespace exists](https://github.com/apache/iceberg/pull/5305#discussion_r927178316).
 e.g. `my_catalog.system.truncate(width, value)` or `system.truncate(6, column)`

## Truncate

The truncate function also allows for a dynamic width or the width to come from a column - though typically the width will likely be static for one given call as it's mostly intended to be used to match partition transforms (specifically with joins or on non-partition columns to create a new column in the data without needing to partition on it).

This PR refactors the definition of the transform functions into a utility class where needed so that Spark’s magic functions can call them via the static `invoke` function and not duplicate logic. This allows Spark to include the functions in codegen.

### Special Considerations for Using Function Catalog Efficiently via Magic Functions and Code Gen

The requirements for magic functions to be used with codegen include that:
1. `invoke` is a static function
4. `invoke` takes in the primitive types / native Spark types corresponding to each of Spark's input DataTypes (e.g. int for IntegerType and UTF8String for StringType).

Further documentation on the magic functions is found [here in the ScalarFunction JavaDoc](https://spark.apache.org/docs/latest/api/java/index.html?org/apache/spark/sql/connector/catalog/functions/ScalarFunction.html)

This partially closes https://github.com/apache/iceberg/issues/5349

### CODE REVIEW

**Comment by kbendick** [comment]
After this, I'll add `bucket` and `zorder` as well.

This is to facilitate the usage of the various transforms from PySpark as well as SQL.

Additionally, having a `zorder` function will make it possible for people to pre-sort their data on input, vs having to zorder sort it when running a data compaction job.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Implmenetation of {@link UnboundFunction} that matches the {@code truncate} transformation.
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:46]: Typo: implementation.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java [hunk 2]
```diff
@@ -46,4 +54,56 @@ public Procedure loadProcedure(Identifier ident) throws NoSuchProcedureException
...
+   *     The namespace is not specified,
+   *     i.e. {@code SELECT truncate(1, 4)} or {@code SELECT my_catalog.truncate(1, 4)}
+   *   </li>
+   *   <li>
+   *     The implicit <b>system</b> namespace is used, to mirror call procedure syntax,
+   *     i.e. {@code SELECT system.truncate(1, 4)} or {@code SELECT my_catalog.system.truncate(1, 4)}
+   *   </li>
+   *   <li>
+   *     A namespace that exists within the catalog is referenced,
+   *     i.e. {@code SELECT ns.truncate(1, 4)} or {@code SELECT my_catalog.ns.truncate(1, 4)}
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java:97]: Similar to my feedback on the Flink PR, this should not expose functions in every database. This should choose a database where Iceberg functions will be exposed. I recommend `system` since that's what we use for stored procedures.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+ * The various implementations are registered with the {@link org.apache.iceberg.spark.SparkCatalog}
+ * such that the function can be used as {@code truncate(col, width)} or {@code truncate(col, 2)}.
+ */
+public class TruncateFunction implements UnboundFunction {
+
+  public TruncateFunction() {
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:52]: Isn't this the default? You should be able to omit empty public constructors.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/SparkFunctions.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  private static final Map<String, UnboundFunction> FUNCTIONS = initFunctions();
+
+  private SparkFunctions() {
+  }
+
+  // The caller needs to validate that the namespace exists and is an Iceberg namespace when using these results.
+  //
+  // Functions that are added to all Iceberg catalogs and their associated namespaces get registered
+  // under Namespace.empty(), allowing the calling code to add the namespace.
+  public static String[] list() {
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/SparkFunctions.java:38]: Why does this return an array?

If it were a list, this could produce an `ImmutableList` and make that a static field, rather than getting the key set and converting it to an array every call.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Implementation of {@link UnboundFunction} that matches the {@code truncate} transformation.
+ * The various implementations are registered with the {@link org.apache.iceberg.spark.SparkCatalog}
+ * such that the function can be used as {@code truncate(col, width)} or {@code truncate(col, 2)}.
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:48]: I don't think this is accurate. Isn't this one unbound function returned by name, not "various implementations"?

Also, the function arguments aren't handled by the catalog lookup. Those are handled by `bind`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format("Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField fieldToTruncate = inputType.apply(0);
+    StructField truncationWidth = inputType.apply(1);
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:81]: The Spark [built-in `bucket` function](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L4976-L4978) makes the width the first argument. I think we should probably stick with that convention for `bucket` and `truncate`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    @Override
+    public boolean isResultNullable() {
+      return true;
+    }
+
+    @Override
+    public boolean isDeterministic() {
+      return true;
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:131]: `true` is the default for both `isDeterministic` and `isResultNullable`, so these implementations shouldn't be necessary.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+    DataType toTruncateSparkType = fieldToTruncate.dataType();
+    Type sourceIcebergType = SparkSchemaUtil.convert(fieldToTruncate.dataType());
+
+    validateSourceTypeUsingSparkTypes(fieldToTruncate.dataType());
+    validateTruncationWidthType(truncationWidth.dataType());
+
+    switch (sourceIcebergType.typeId()) {
+      case INTEGER:
+        return new TruncateInteger();
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:91]: I think this is going to cause problems. Both `tinyint` and `smallint` are mapped to Iceberg's `int`. It's fine to use the same bound function for all of them, but right now `TruncateInteger` assumes that the input type is an int and calls `getInt`, which is going to produce incorrect values when reading an `UnsafeRow` for the smaller types. It would also fail when using a `GenericRow`.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+  public TestSparkTruncateFunction(String catalogName, String implementation, Map<String, String> config) {
+    super(catalogName, implementation, config);
+    this.fullNamespace = ("spark_catalog".equals(catalogName) ? "" : catalogName + ".") + NS;
+  }
+
+  @Before
+  public void createNamespace() {
+    sql("CREATE NAMESPACE IF NOT EXISTS %s", fullNamespace);
+    sql("USE %s", fullNamespace);
```
Comment (rdblue) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:49]: I don't think you need a namespace if you always reference `system.truncate`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 8 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:91]: When reading the documentation for `ScalarFunction`, it seemed like Spark wouls perform the required cast before invoking the magic function.

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:91]: https://spark.apache.org/docs/latest/api/java/index.html?org/apache/spark/sql/connector/catalog/functions/ScalarFunction.html

I might have that wrong though.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  public static class TruncateInteger extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
+      }
+
+      return value - (((value % width) + width) % width);
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:142]: I'd prefer to use the actual truncate function. I talked with Dan and heard that the motivation for this was that width might change and we don't want to create a new instance of the transform every time. I think it's good to catch that possibility, but there are other options that avoid copying the transform implementation.

The option I like best is to assume that the width is not going to change across calls and check that width is the same as the last row's width:

```java
  private int lastWidth = -1;
  private Transform<Integer, Integer> transform = null;
  public static int invoke(int width, int value) {
    ...

    if (lastWidth != width) {
      this.lastWidth = width;
      this.transform = Transforms.truncate(IntegerType.get(), width);
    }

    return transform.apply(value);
  }
```

For constant values, this will work great. There's some overhead for changing widths, but I don't think that's going to be a common case since it makes very little sense.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 8 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:91]: Oh actually you're right. I think I need to implement one for `tinyint` and `smallint` if we want to keep them using the magic funcitons.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+    public boolean isDeterministic() {
+      return true;
+    }
+  }
+
+  public static class TruncateInteger extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:139]: I probably wouldn't bother with this check.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 10 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:142]: Agreed that it's not likely to be a common case. That's not really the intended purpose of the function and I don't care for the code duplication.

Will go with that approach then. Thank you!

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    @Override
+    public String canonicalName() {
+      return "org.apache.iceberg.spark.functions.truncate(int,int)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
+      return invoke(input.getInt(0), input.getInt(1));
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:162]: Does this need to handle null `value` args?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.StringType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "org.apache.iceberg.spark.functions.truncate(string,int)";
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:223]: I'd probably make this `iceberg.truncate[width](string)`, but it's minor.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 10 above)
Comment (rdblue) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:142]: Actually, I'm reconsidering this. The option of having a static function that performs this would allow us to avoid boxing/unboxing and the lookup entirely. That's probably the better thing to do.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 12 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:162]: It does, but since it has the magic function, it should already return `null` when it receives any `null` input.

I'm going to add counters to make sure the magic functions are the only ones being called though and then possibly remove `produceResult`.

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:162]: For now I'm updating it to handle `null` though 👍

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java (hunk 2 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java]: So the only problem with this is that within SQL (or even the dataframe API), the `SparkSessionCatalog` checks that the namespace exists.

So calling `system.truncate` results in:

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java]: So the only problem with this is that within SQL (or even the dataframe API), the `SparkSessionCatalog` checks that the namespace exists.

So calling `system.truncate` results in
```
org.apache.spark.sql.AnalysisException: Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.
```

I'm seeing if I can find a way around this, but it is part of the SessionCatalog's code.

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java]: Full stack trace:

```
Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.; line 1 pos 7
org.apache.spark.sql.AnalysisException: Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.; line 1 pos 7
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.failFunctionLookup(SessionCatalog.scala:1561)
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.resolvePersistentFunctionInternal(SessionCatalog.scala:1704)
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.resolvePersistentFunction(SessionCatalog.scala:1673)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$.org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveFunctions$$resolveV1Function(Analyzer.scala:2172)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$$anonfun$apply$25$$anonfun$applyOrElse$103.$anonfun$applyOrElse$109(Analyzer.scala:2122)
	at app//scala.Option.getOrElse(Option.scala:189)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$$anonfun$apply$25$$anonfun$applyOrElse$103.$anonfun$applyOrElse$108(Analyzer.scala:2119)
```

Link to the code in the analyzer that's being hit: https://github.com/apache/spark/blob/f74867bddfbcdd4d08076db36851e88b15e66556/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2118-L2126

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/SparkFunctions.java (hunk 4 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/SparkFunctions.java]: Made it an immutable list and converted to a static field.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 5 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Updated, though the more I think about it, the more I think I might throw this out.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 6 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Changed width to be the first argument.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 7 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Removed the unnecessary overrides.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java (hunk 9 above)
Comment (kbendick) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java]: That's correct. I'm still working on how to handle the `system` namespace.

I think probably we just register it as a special namespace within the `SparkSessionCatalog`? The analyzer specifically checks if the namspace exists when using the session catalog.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 11 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:126]: Should we just let the code fail? For some calls, it seems the code would do things that don't necessarily work (based on the existing transforms).

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 13 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Yeah I like that more. Updated.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 8 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:90]: Looking into this further, it seems that Spark will add a cast for us to Integer if we allow `smallint` or `tinyint` but only define a function for Integer type.

When calling `EXPLAIN EXTENDED SELECT system.truncate(2, -1S)`, there is a `(CAST -1 as int)` in the logical plan:

```
== Parsed Logical Plan ==
'Project [unresolvedalias('testhadoop.system.truncate(2, -1), None)]
+- OneRowRelation

== Analyzed Logical Plan ==
staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateInteger, IntegerType, invoke, 2, -1, IntegerType, IntegerType, false, true, true): int
Project [staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateInteger, IntegerType, invoke, 2, cast(-1 as int), IntegerType, IntegerType, false, true, true) AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateInteger, IntegerType, invoke, 2, -1, IntegerType, IntegerType, false, true, true)#98]
+- OneRowRelation

== Optimized Logical Plan ==
Project [-2 AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateInteger, IntegerType, invoke, 2, -1, IntegerType, IntegerType, false, true, true)#98]
+- OneRowRelation

== Physical Plan ==
*(1) Project [-2 AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateInteger, IntegerType, invoke, 2, -1, IntegerType, IntegerType, false, true, true)#98]
+- *(1) Scan OneRowRelation[]
```

However, the resulting row type will still be `IntegerType`, so we should consider still defining them.

But I'm ok with allowing the cast for the width field from tinyint / smallint -> int as that doesn't affect the resulting input type and the static function is still called.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 11 above)
Comment (rdblue) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:130]: It won't fail because these are primitives.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java (hunk 2 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java]: We spoke offline and we're just giong to hold off on the spark session catalog for now.

Most likely in a follow up we will add a namespace `system` that will then be resolvable (as these issues go away when the namespace is resolvable). That matches other databases, as well as the experience of using the `ProcedureCatalog`.

I've removed or `@Ignore`d the tests against `SparkSessionCatalog` in this PR.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 8 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: I added `tinyint` and `smallint` support, so the return type would be correct.

For the width, as shown in the explain, Spark will cast it (and there is logic to ensure that width is only one of `ByteType` / `ShortType` / `IntType` aka `tinytint` / `smallint` / `int`).

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java (hunk 9 above)
Comment (kbendick) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java]: Removed the namespace from here as it's not needed.

As mentioned here https://github.com/apache/iceberg/pull/5305#discussion_r927178316, we will hold off on `SparkSessionCatalog` and then follow up with handling the need for `system` namespace in spark session catalog in another PR.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 11 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:220]: Buyt the width could be negative. The range for Integer is signed integer range. That's my concern (not so much that we'll get null for the width).

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 12 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Have updated all to handle null. Also have tests that check for calling truncate on `null` of each specific type (via casting).

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 11 above)
Comment (rdblue) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:218]: Yeah, if you use weird values, you get weird results.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+  }
+
+  public abstract static class TruncateBase<T> implements ScalarFunction<T> {
+    @Override
+    public String name() {
+      return "truncate";
+    }
+  }
+
+  // tinyint
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:138]: Nit: this comment duplicates what is already in the class name.

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:138]: Thanks. I noticed that now that I renamed them. Moving the functions to be in a utility class and then this will be good to take a look at =)

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java [hunk 15]
```diff
@@ -46,4 +52,30 @@ public Procedure loadProcedure(Identifier ident) throws NoSuchProcedureException
...
+
+  @Override
+  public Identifier[] listFunctions(String[] namespace) throws NoSuchNamespaceException {
+    if (namespace.length == 1 && namespace[0].equalsIgnoreCase("system")) {
+      return SparkFunctions.list().stream()
+          .map(name -> Identifier.of(namespace, name))
+          .toArray(Identifier[]::new);
+    }
+
+    throw new NoSuchNamespaceException(namespace);
```
Comment (rdblue) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java:64]: This should return an empty list if the database exists.

Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java:68]: Fixed

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 10 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: I moved the truncation functions into `TruncateUtil` in api.

Some things already had utility functions handling their truncation, such as `UnicodeUtil.truncateString` so I didn't add functions for that.

The code in `TruncateUtil` explicitly assumes that the input has been validated, to allow for more efficient codegen or for skipping validation if it's already been done.
We pre-validate:
1. The Truncate `Transform` validates the `width` in their constructor so it doesn't need to be done again.
2. The Spark Truncate `BoundFunction` will immediately return `null` for `null` input for primitives.

Not doing duplicate validation was chosen because this code is in the hot path (namely because it gets included in Spark's codegen when used with `FunctionCatalog`).

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 11 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:218]: I'll add a note in the JavaDoc and then remove that (as throwing exceptions in generated code is somewhat of an anti-pattern if possible).

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java [hunk 16]
```diff
@@ -394,9 +395,7 @@ public ByteBuffer apply(ByteBuffer value) {
         return null;
       }
 
-      ByteBuffer ret = value.duplicate();
-      ret.limit(Math.min(value.limit(), value.position() + length));
-      return ret;
+      return TruncateUtil.truncateByteBuffer(length, value);
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/transforms/Truncate.java:398]: There's `BinaryUtil#truncateBinary(ByteBuffer, int)`, but it doesn't exactly have the same code as is used here and I think the logic is different in that it allows for 0 length.

I can try to merge the two implementations in a follow up if we want.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java (hunk 10 above)
Comment (kbendick) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java]: Extracted out a `TruncateUtil`, at least for the calls to be shared between `Truncate` transform and the generated code. As both `invoke` and the `TruncateUtil` are both static functions, will the Spark generated code remove the nested static function call?

**Comment by kbendick** [comment]
@rdblue PTAL. I was going to tag other people as well.

**Comment by kbendick** [comment]
@rdblue Do you think I should break this into a 2 PRs?

**Comment by kbendick** [comment]
Looking at the storage partition joins, it looks like the function `bucket` will need to be resolvable on the empty namespace. I can update that here or a follow up:

https://github.com/apache/spark/blob/e9eb28e27d10497c8b36774609823f4bbd2c8500/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala#L84-L100

**Comment by aokolnychyi** [comment]
I'd love to take a look as well. I should have some time in a day. We also had some progress on bucketed joins internally.

**Comment by kbendick** [comment]
> I'd love to take a look as well. I should have some time in a day. We also had some progress on bucketed joins internally.

Thanks Anton. Was going to tag you today now that it's cleaned up. Also cc @huaxingao @flyrain @nastra @Fokko

**Comment by kbendick** [comment]
Right not we're requiring the call be to the `system` namespace, but the storage partitioned join implementation looks for a function called `bucket` in the `FunctionCatalog` using an empty array for the namespace per [this diff in the original Storage Partitioned Joins PR in `V2ExpressionUtils#toCatalystTransform`.](https://github.com/apache/spark/commit/20ffbf7b308c3dc90a49dbdcb8d7b972eeb53bc4#diff-715d0c2d59a4ddb8a4b5952c1b05be9f035b6d9b0d9670c70b58989dc722b252R79-R95)

So we should probably allow the empty namespace to resolve functions as well.

**Comment by kbendick** [comment]
Link to the code that resolves out `bucket` implementation in case the link above doesn't resolve: https://github.com/apache/spark/blob/47f0303944abb11d3018186bc125113772eff8ef/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala#L84-L100

**Comment by kbendick** [comment]
Because this PR is so big, I'm going to separate out the `FunctionCatalog` implementation from `Truncate`.

I'm going to add a very simple function to be able to test it but keep the code to review a lot smaller. 👍

**Comment by kbendick** [comment]
I've opened https://github.com/apache/iceberg/pull/5377 to cover _just_ the `FunctionCatalog`.

This PR is too big, and this way we can focus on just the `FunctionCatalog` business without having to worry about the details of `truncate`.

I've added an `iceberg_version` function in the other PR to assist with testing.

**Comment by kbendick** [comment]
This PR is closed in favor of https://github.com/apache/iceberg/pull/5377 and https://github.com/apache/iceberg/pull/5411.

I'll open a PR for `Truncate` and link it shortly.
